Expose cloudserver metrics on separate port and /metrics path

francoisferrand opened the pull request:
This reduces the attack surface, and ensure we do not leak (metrics) information via the `_/monitoring/metrics` path.

Issue: CLDSRV-51

Commit:
feature: CLDSRV-51 - switch case for prometheus object updates

Commit:
feature: CLDSRV-51 - spawn a new HTTP server to serve metrics

francoisferrand committed:
Proper routing of metrics requests

Issue: CLDSRV-51

francoisferrand committed:
Expose metrics port in Docker image

Issue: CLDSRV-51

francoisferrand committed:
Make metrics endpoint configurable

Issue: CLDSRV-51

francoisferrand committed:
Ensure to close all servers on shutdown

Issue: CLDSRV-51

bert-e commented:
# Hello francoisferrand,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://eve.devsca.com/github/scality/cloudserver/bert-e/doc/user).





*Status report is not available.*

bert-e commented:
# Incorrect fix version



The `Fix Version/s` in issue CLDSRV-51 contains:


* *None*



Considering where you are trying to merge, I ignored possible hotfix versions and I expected to find:



* `8.3.0`


Please check the `Fix Version/s` of CLDSRV-51, or the target
branch of this pull request.

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* 2 peers

lavigne958 commented:
~you don't have any ticket number in your PR/commits :thinking: 
you include commits from CLDSRV-51 from @jbertran 
should this PR need to cleaning first ? wait for the CLDSRV-51 to pass first ?~
got it, sorry

alexanderchan-scality commented on lib/utilities/monitoringHandler.js:
```diff
@@ -167,7 +179,7 @@ function monitoringHandler(clientIP, req, res, log) {
         return monitoringEndHandler(res, errors.MethodNotAllowed);
     }
-    const monitoring = (req.url === '/_/monitoring/metrics');
+    const monitoring = (req.url === '/_/metrics');
```
small change: 
the constant defined here can be used.
https://github.com/scality/Arsenal/blob/development/8.1/lib/network/probe/ProbeServer.js#L7

francoisferrand replied:
This is pending discussion on Confluence: https://scality.atlassian.net/wiki/spaces/OS/pages/1819148292/Strategy+for+exposing+metrics (where we consider using a simpler `/metrics` path, since we are already moved to a different port)

Also, importing from Arsenal does not seem look so appropriate with the current Arsenal implementation : this constant is present indeed, but deep inside the code of ProbeServer, and looks more like an implementation detail than an "established" constant we should rely on.

miniscruff replied:
The constant was meant to be relied upon regardless of usage to probe server, as probe server itself is optional for metrics/probes as your service may already have an established server. It is of course not "established" because it is new and only used a couple of times in backbeat as that is where it was initialized built for.

francoisferrand replied:
in that case, should we not move the constant in arsenal's `constants` ?

miniscruff replied:
Uh maybe? Our use of constant.js files is a little confusing to me and I tend to go with more contextual constants rather than one mega file with everything under the sun. If you were discouraged to use the value because it was not placed in a constants file I guess we should move it. I got discouraged when I would use backbeat and there are multiple constants ( one in lib/ and one in each extension ).

francoisferrand committed:
Merge remote-tracking branch 'origin/development/8.3' into feature/CLDSRV-51-new-http-server-for-metrics

bert-e commented:
# Conflict




There is a conflict between your branch `feature/CLDSRV-51-new-http-server-for-metrics` and the
destination branch `development/8.3`.

Please resolve the conflict on **the feature branch** (`feature/CLDSRV-51-new-http-server-for-metrics`).

```sh
git fetch && \
git checkout origin/feature/CLDSRV-51-new-http-server-for-metrics && \
git merge origin/development/8.3
```

Resolve merge conflicts and commit

```sh
git push origin HEAD:feature/CLDSRV-51-new-http-server-for-metrics
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* 2 peers

francoisferrand committed:
Use /metrics path

As defined in https://scality.atlassian.net/wiki/spaces/OS/pages/1819148292/Strategy+for+exposing+metrics

Issue: CLDSRV-51

francoisferrand committed:
Use port 8002 for cloudserver metrics

Issue: CLDSRV-51

francoisferrand commented:
/approve

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/8.3`



The following branches will **NOT** be impacted:

* `development/6.4`
* `development/7.10`
* `development/7.4`
* `development/8.2`
* `stabilization/7.10.1`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **approve**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/8.3`



The following branches have **NOT** changed:

* `development/6.4`
* `development/7.10`
* `development/7.4`
* `development/8.2`
* `stabilization/7.10.1`



Please check the status of the associated issue CLDSRV-51.

Goodbye francoisferrand.
